### PR TITLE
amp-render bento component

### PR DESCRIPTION
**Pull request by dmanek**
Partial for #32477

Features addressed in this PR:
- [X] fetch remote JSON specified via src
- [X] fetch JSON from amp-state
- [X] unit tests
- [X] storybook tests

### CODE REVIEW

File: extensions/amp-render/1.0/amp-render.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    //   isExperimentOn(this.win, 'amp-render'),
+    //   'Experiment "amp-render" is not turned on.'
+    // );
+    return super.isLayoutSupported(layout);
+  }
+
+  /** @override */
+  init() {
+    return dict({
+      'fetchFn': this.getFetchFn_(),
```
Comment (caroqliu) [inline extensions/amp-render/1.0/amp-render.js:55]:
```suggestion
      'fetchFn': this.getFetchFn_,
```

Comment (dmanek) [reply extensions/amp-render/1.0/amp-render.js:55]: I think you misunderstood. `this.getFetchFn` looks at src to determine if its amp-state, amp-script, etc. & returns the appropriate function to perform the fetch.

Comment (caroqliu) [reply extensions/amp-render/1.0/amp-render.js:55]: Ooooh you're right, thanks for clarifying. Sorry I misread!

Comment (caroqliu) [reply extensions/amp-render/1.0/amp-render.js:55]: Ooooh you're right, thanks! Will you have to add a mutation observer to `src` to make sure you switch the `fetchFn` if the src changes?

Comment (samouri) [reply extensions/amp-render/1.0/amp-render.js:55]: are we moving off of mutatedAttributeCallback?

Comment (caroqliu) [reply extensions/amp-render/1.0/amp-render.js:55]: It should just work with `mutationObserverCallback`, since the bound `[src]` operation will mutate the `src` attribute with the updated value.

I don't think we can rely on just `BaseElement['props']`'s configuration of `src` (which already puts a mutation observer on those attrs) since we would want to be able to mutate a different prop `fetchFn` based on `src` mutating... but I wonder if we can do something like the following, assuming we can make `getFetchFn` a top level fn that takes an element instead of inside `AmpRender`:
```
BaseElement['props'] = {
  'src': { attr: 'src' },
  'fetchFn': { attrs: ['src'], parseAttrs: (e) => getFetchFn(e)
}
```

Comment (dmanek) [reply extensions/amp-render/1.0/amp-render.js:54]: If such a thing is possible, we can look into moving `getFetchFn` at the top level.

File: extensions/amp-render/1.0/amp-render.js [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  /**
+   * Returns true if element's src points to amp-state.
+   *
+   * @param {string} src
+   * @return {boolean}
+   * @private
+   */
+  isAmpStateSrc_(src) {
```
Comment (caroqliu) [inline extensions/amp-render/1.0/amp-render.js:65]: This can be moved out of the class to top level

File: extensions/amp-render/1.0/amp-render.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  /**
+   * Returns true if element's src points to an amp-script function.
+   *
+   * @param {string} src
+   * @return {boolean}
+   * @private
+   */
+  isAmpScriptSrc_(src) {
```
Comment (caroqliu) [inline extensions/amp-render/1.0/amp-render.js:76]: Same as above

File: extensions/amp-render/1.0/amp-render.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+   * @private
+   */
+  getFetchFn_() {
+    const src = this.element.getAttribute('src');
+    if (this.isAmpStateSrc_(src)) {
+      return this.getAmpStateJson.bind(null, this.element);
+    }
+    if (this.isAmpScriptSrc_(src)) {
+      // TODO(dmanek): implement this
+      return;
```
Comment (caroqliu) [inline extensions/amp-render/1.0/amp-render.js:94]: Let's have this return noop in the mean time so it still returns a function type
```suggestion
      return () => {};
```

File: extensions/amp-render/1.0/amp-render.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    }
+    return batchFetchJsonFor.bind(null, this.getAmpDoc(), this.element);
+  }
+
+  /** @override */
+  checkPropsPostMutations() {
```
Comment (caroqliu) [inline extensions/amp-render/1.0/amp-render.js:100]: This looks identical to the ones in `amp-date-display` and `amp-date-countdown` -- can you add a note (not blocking for this PR) to see if it's worth making it shareable?

File: extensions/amp-render/1.0/amp-render.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+          );
+        });
+      } else {
+        this.mutateProps(dict({'render': null}));
+      }
+    }
+  }
+
+  /** @override */
+  isReady(props) {
```
Comment (caroqliu) [inline extensions/amp-render/1.0/amp-render.js:131]: Same as above.

File: extensions/amp-render/1.0/base-element.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+export class BaseElement extends PreactBaseElement {}
+
+/** @override */
+BaseElement['Component'] = Render;
+
+/** @override */
+BaseElement['props'] = {
+  'src': {attr: 'src'},
+  'fetchFn': {attr: 'fetch-fn'},
```
Comment (caroqliu) [inline extensions/amp-render/1.0/base-element.js:28]: This can be removed since we can't pass functions via DOM attributes

File: extensions/amp-render/1.0/component.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import * as Preact from '../../../src/preact';
+import {Wrapper, useRenderer} from '../../../src/preact/component';
+import {useEffect, useState} from '../../../src/preact';
+import {useResourcesNotify} from '../../../src/preact/utils';
+
+/**
+ * @param {!RenderDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function Render({src, fetchFn, render, ...rest}) {
```
Comment (caroqliu) [inline extensions/amp-render/1.0/component.js:26]: nit: can `fetchFn` just be `fetch`?

Also it's not necessarily always fetching i.e. with static data like amp-state. Would `getData` or `getDataAsync` make sense instead?

Comment (caroqliu) [inline extensions/amp-render/1.0/component.js:26]: Should `fetchFn` and `render` have default values? What if `src` is `undefined` or otherwise invalid?

File: extensions/amp-render/validator-amp-render.protoascii [hunk 9]
```diff
@@ -0,0 +1,42 @@
+#
```
Comment (caroqliu) [inline extensions/amp-render/validator-amp-render.protoascii:1]: Let's not put the validations in this PR -- this should be a separate PR when we're closer to launch.

File: extensions/amp-render/1.0/test/test-amp-render.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  (env) => {
+    let win, doc, element;
+
+    beforeEach(async function () {
+      win = env.win;
+      doc = win.document;
+      toggleExperiment(win, 'amp-render', true, true);
+    });
+
+    it('renders', async () => {
```
Comment (caroqliu) [inline extensions/amp-render/1.0/test/test-amp-render.js:37]: Worth adding tests for amp-state data and remote fetched data, as well as unit tests for your helper fns.

File: extensions/amp-render/1.0/test/test-component.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+  it('should render as a div by default', () => {
+    const fetchStub = sandbox.stub();
+    fetchStub.resolves(response);
+
+    const render = (data) => JSON.stringify(data);
+
+    const props = {
+      fetchFn: fetchStub,
+      src: 'https://example.com',
+      render,
```
Comment (caroqliu) [inline extensions/amp-render/1.0/test/test-component.js:41]: Let's also test behaviors for invalid or missing props

**Review by caroqliu — CHANGES_REQUESTED** [review]

File: extensions/amp-render/1.0/component.js (hunk 8 above)
Comment (dmanek) [reply extensions/amp-render/1.0/component.js]: Added default values for both. I still need to add url replacement logic for `src`, so additional validation will be added then (added TODO for it).

File: extensions/amp-render/1.0/amp-render.js [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+   * src="amp-state:json.path".
+   *
+   * TODO: this implementation is identical to one in amp-list. Move it
+   * to a common file and import it.
+   *
+   * @param {!AmpElement} element
+   * @return {Promise<!JsonObject>}
+   */
+  getAmpStateJson(element) {
+    const src = element.getAttribute('src');
```
Comment (caroqliu) [inline extensions/amp-render/1.0/amp-render.js:152]: Guard for if `src` is `null`, since we perform `slice` on it later. Also, let's make `src` a parameter, since it is called with that as an argument in `Render`.

File: extensions/amp-render/1.0/storybook/Basic.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  component: Render,
+  decorators: [withKnobs, withA11y],
+};
+
+export const _default = () => {
+  return (
+    <Render
+      src={'http://example.com'}
+      getJson={() => Promise.resolve({name: 'George'})}
+      render={(data) => `Hi ${data.name}!`}
```
Comment (caroqliu) [inline extensions/amp-render/1.0/storybook/Basic.js:33]: Do you mind adding either new example cases or knobs for trying the default fetch/render props when none is provided?

File: extensions/amp-render/1.0/test/test-amp-render.js [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+              "name": "Bill"
+            }
+          </script>
+        </amp-state>
+      `;
+      doc.body.appendChild(ampState);
+
+      element = html`
+        <amp-render
+          src="amp-state:theFood"
```
Comment (caroqliu) [inline extensions/amp-render/1.0/test/test-amp-render.js:72]: Please also add a case where `src` is omitted -- does the component degrade gracefully?

File: extensions/amp-render/1.0/test/test-amp-render.js [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+      const ampRender = await element.getImpl(false);
+
+      const fetchFn = () => {
+        return Promise.resolve({name: 'Joe'});
+      };
+
+      env.sandbox.stub(ampRender, 'getJsonFn_').returns(fetchFn);
+
+      const text = await getRenderedData();
+      expect(text).to.contain('Hello Joe');
```
Comment (caroqliu) [inline extensions/amp-render/1.0/test/test-amp-render.js:111]: Are these using `contain` instead of `equal` due to white space? If so, can we `trim` first to assert `equal`? It shouldn't render anything else.

File: extensions/amp-render/1.0/test/test-component.js [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../../src/preact';
+import {Render} from '../component';
+import {macroTask} from '../../../../testing/yield';
+import {mount} from 'enzyme';
+
+describes.sandboxed('Render 1.0 preact component', {}, (env) => {
```
Comment (caroqliu) [inline extensions/amp-render/1.0/test/test-component.js:22]:
```suggestion
describes.sandboxed('Render 1.0 preact component', {}, () => {
```

**Review by caroqliu — CHANGES_REQUESTED** [review]
Had a few suggestions, looks good overall. 👍  ⚡

File: extensions/amp-render/1.0/amp-render.js [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+  /**
+   * Returns the correct fetch function for amp-state, amp-script or
+   * to fetch remote JSON.
+   *
+   * @return {Function}
+   * @private
+   */
+  getJsonFn_() {
+    const src = this.element.getAttribute('src');
```
[the remark this answers is not in the document]
Comment (dmanek) [reply extensions/amp-render/1.0/amp-render.js:80]: Think it makes sense to guard against `null` in `isAmpStateSrc` & `isAmpScriptSrc` methods as well as here. I've added a TODO to assert that `src` is provided.

File: extensions/amp-render/1.0/test/test-amp-render.js (hunk 15 above)
Comment (dmanek) [reply extensions/amp-render/1.0/test/test-amp-render.js]: `contain` was a remnant of copy/paste. Changed to `equal`.

File: extensions/amp-render/1.0/amp-render.js (hunk 1 above)
Comment (dmanek) [reply extensions/amp-render/1.0/amp-render.js]: This PR is already pretty large. I'll make these changes in a follow up PR.

File: extensions/amp-render/1.0/amp-render.js (hunk 12 above)
Comment (dmanek) [reply extensions/amp-render/1.0/amp-render.js:156]: Added guard clause for when `src` is `null`. I'll add src as param in the follow up PR which will require moving `getAmpStateJson` out of `AmpRender` class  to support src binding

Comment (caroqliu) [reply extensions/amp-render/1.0/amp-render.js:156]: Okay, in that case please add it as a TODO or in a tracking issue so we can keep track of it.

**Review by caroqliu — APPROVED** [review]

**Review by rileyajones — APPROVED** [review]

File: extensions/amp-render/1.0/amp-render.js [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    }
+    if (isAmpScriptSrc(src)) {
+      // TODO(dmanek): implement this
+      return () => {};
+    }
+    return batchFetchJsonFor.bind(null, this.getAmpDoc(), this.element);
+  }
+
+  /**
+   * TODO: this implementation is identical to one in amp-data-display &
```
Comment (samouri) [inline extensions/amp-render/1.0/amp-render.js:96]: amp-date-display

File: extensions/amp-render/1.0/amp-render.js [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+   * amp-date-countdown. Move it to a common file and import it.
+   *
+   * @override
+   */
+  checkPropsPostMutations() {
+    const templates =
+      this.templates_ ||
+      (this.templates_ = Services.templatesForDoc(this.element));
+    const template = templates.maybeFindTemplate(this.element);
+    if (template != this.template_) {
```
Comment (samouri) [inline extensions/amp-render/1.0/amp-render.js:106]: super nitty: if you invert the condition via `if (template === this.template_)`, then you can early return and unindent the rest of the fn.

File: extensions/amp-render/1.0/amp-render.js [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+          );
+        });
+      } else {
+        this.mutateProps(dict({'render': null}));
+      }
+    }
+  }
+
+  /**
+   * TODO: this implementation is identical to one in amp-data-display &
```
Comment (samouri) [inline extensions/amp-render/1.0/amp-render.js:132]: same nit re. amp-data --> amp-date

File: extensions/amp-render/1.0/amp-render.js [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+   * amp-date-countdown. Move it to a common file and import it.
+   *
+   * @override
+   */
+  isReady(props) {
+    if (this.template_ && !('render' in props)) {
```
Comment (samouri) [inline extensions/amp-render/1.0/amp-render.js:138]: nit: function body can be reduced to

```js
// If a template is specified, then it must be available.
return !this.template_ || ('render' in props)
```

File: extensions/amp-render/1.0/amp-render.js [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+      // The template is specified, but not available yet.
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Gets the json an amp-list that has an "amp-state:" uri. For example,
```
Comment (samouri) [inline extensions/amp-render/1.0/amp-render.js:146]: Can you update the comment to reference `amp-render` instead of `amp-list`? Or make comment more generic when moved to a common file.

File: extensions/amp-render/1.0/component.js [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+ * @param {!JsonObject} data
+ * @return {string}
+ */
+const DEFAULT_RENDER = (data) => JSON.stringify(data);
+
+/**
+ * @param {string} url
+ * @return {!Promise<!JsonObject>}
+ */
+const DEFAULT_FETCH = (url) => {
```
Comment (samouri) [inline extensions/amp-render/1.0/component.js:32]: Lets rename this to `DEFAULT_GET_JSON`

**Review by samouri — COMMENTED** [review]
Great start to the component!
Sorry for reviewing post-merge, but it was only out of draft for 2 hours 😬

**Comment by dmanek** [comment]
My apologies, jumped the gun there. Thanks for your comments. I've addressed them in WIP PR #33426 for making the src bindable.

File: extensions/amp-render/1.0/component.js [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+
+  useEffect(() => {
+    // TODO(dmanek): Add additional validation for src
+    // when adding url replacement logic.
+    if (!src) {
+      return;
+    }
+
+    getJson(src).then((data) => {
+      setData(data);
```
Comment (jridgewell) [inline extensions/amp-render/1.0/component.js:58]: There's a  hidden race condition here where the `getJson` request takes to long and we've already rerendered the component. We should be checking if this effect has been cancelled:

```js
useEffect(() => {
  let cancelled = false;
  getJson(src).then(data => {
    if (!cancelled) {
      setData(data);
    }
  });
  return () => {
    cancelled = true;
  };
});

Comment (samouri) [reply extensions/amp-render/1.0/component.js:58]: Good find! This isn't purely theoretical either. We had to fix it for amp-list as well: https://github.com/ampproject/amphtml/pull/25964

Comment (dmanek) [reply extensions/amp-render/1.0/component.js:58]: Thanks for the comment.  I've addressed it in WIP PR #33426.
